sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergIOTranslation.java

```diff
@@ -0,0 +1,241 @@
+        }
+        return builder.build();
+      } catch (InvalidClassException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  @AutoService(TransformPayloadTranslatorRegistrar.class)
+  public static class ReadRegistrar implements TransformPayloadTranslatorRegistrar {
+    @Override
+    @SuppressWarnings({
+      "rawtypes",
+    })
+    public Map<? extends Class<? extends PTransform>, ? extends TransformPayloadTranslator>
+        getTransformPayloadTranslators() {
+      return ImmutableMap.<Class<? extends PTransform>, TransformPayloadTranslator>builder()
+          .put(AutoValue_IcebergIO_ReadRows.class, new IcebergIOReadTranslator())
+          .build();
+    }
+  }
+
+  static class IcebergIOWriteTranslator implements TransformPayloadTranslator<WriteRows> {
+
+    static final Schema WRITE_SCHEMA =
+        Schema.builder()
+            .addByteArrayField("catalog_config")
+            .addNullableStringField("table_identifier")
+            .addNullableByteArrayField("dynamic_destinations")
+            .build();
+
+    public static final String ICEBERG_WRITE_TRANSFORM_URN =
+        "beam:transform:org.apache.beam:iceberg_write:v1";
+
+    @Override
+    public String getUrn() {
+      return ICEBERG_WRITE_TRANSFORM_URN;
+    }
+
+    @Override
+    public @Nullable FunctionSpec translate(
+        AppliedPTransform<?, ?, WriteRows> application, SdkComponents components)
+        throws IOException {
+      SchemaApi.Schema expansionSchema = SchemaTranslation.schemaToProto(WRITE_SCHEMA, true);
+      Row configRow = toConfigRow(application.getTransform());
+      ByteArrayOutputStream os = new ByteArrayOutputStream();
+      RowCoder.of(WRITE_SCHEMA).encode(configRow, os);
+
+      return FunctionSpec.newBuilder()
+          .setUrn(getUrn())
+          .setPayload(
+              SchemaAwareTransformPayload.newBuilder()
+                  .setExpansionSchema(expansionSchema)
+                  .setExpansionPayload(ByteString.copyFrom(os.toByteArray()))
+                  .build()
+                  .toByteString())
+          .build();
+    }
+
+    @Override
+    public Row toConfigRow(WriteRows transform) {
+
+      Map<String, Object> fieldValues = new HashMap<>();
+
+      if (transform.getCatalogConfig() != null) {
+        fieldValues.put("catalog_config", toByteArray(transform.getCatalogConfig()));
+      }
+      if (transform.getTableIdentifier() != null) {
+        TableIdentifier identifier = transform.getTableIdentifier();
+        List<String> identifierParts =
+            Arrays.stream(identifier.namespace().levels()).collect(Collectors.toList());
+        identifierParts.add(identifier.name());
+        fieldValues.put("table_identifier", String.join(".", identifierParts));
+      }
+      if (transform.getDynamicDestinations() != null) {
+        fieldValues.put("dynamic_destinations", toByteArray(transform.getDynamicDestinations()));
```

Comment:
As discussed offline, we should consider implementing `Externalizable` for the complex objects serialized here (`CatalogConfig` and `DynamicDestinations`).

Reply:
Done